allennlp/modules/token_embedders/pretrained_transformer_embedder.py

```diff
@@ -23,7 +34,64 @@ def get_output_dim(self):
         return self.output_dim
 
     def forward(
-        self, token_ids: torch.LongTensor, mask: torch.LongTensor
+        self,
+        token_ids: torch.LongTensor,
+        mask: torch.LongTensor,
+        wordpiece_mask: torch.LongTensor = None,
+        offsets: torch.LongTensor = None,
     ) -> torch.Tensor:  # type: ignore
+        """
+        # Parameters
+
+        token_ids: torch.LongTensor
+            Shape: [batch_size, num_wordpieces].
+        mask: torch.LongTensor
+            Shape: [batch_size, num_orig_tokens if self._intra_word_tokenized else num_wordpieces].
+        wordpiece_mask: torch.LongTensor
+            Shape: [batch_size, num_wordpieces].
+        offsets: torch.LongTensor
+            Shape: [batch_size, num_orig_tokens, 2].
+            Maps indices for the original tokens, i.e. those given as input to the indexer,
+            to a span in token_ids. `token_ids[i][offsets[i][j][0]:offsets[i][j][1] + 1]`
+            corresponds to the original j-th token from the i-th batch.
+
+        # Returns:
+        Shape: [
+            batch_size,
+            num_orig_tokens if self._intra_word_tokenized else num_wordpieces,
+            embedding_size,
+        ].
+        """
+        if self._intra_word_tokenized:
+            if wordpiece_mask is None or offsets is None:
+                raise ValueError(
+                    "`wordpiece_mask` and `offsets` must be set if `intra_word_tokenized == True`."
+                )
+
+        device = token_ids.device
+
+        transformer_mask = wordpiece_mask if self._intra_word_tokenized else mask
+        # Shape: [batch_size, num_wordpieces, embedding_size].
+        embeddings = self.transformer_model(input_ids=token_ids, attention_mask=transformer_mask)[0]
+
+        if self._intra_word_tokenized:
+            batch_size, _, embedding_size = embeddings.size()
+            num_orig_tokens = offsets.size(1)
+
+            orig_embeddings = torch.FloatTensor(
+                batch_size, num_orig_tokens, embedding_size
+            ).to(device)
+            for batch_idx in range(batch_size):  # TODO: do we have to use loops?
```

Comment:
I'm going to wait to comment on this TODO until I see it split into a separate file.